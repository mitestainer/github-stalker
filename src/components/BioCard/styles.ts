import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    border-radius: ${theme.borderRadius};
    background-color: ${theme.colors.white};
    border: 1px solid ${theme.colors.purple};
    padding: ${theme.gutter};
    display: grid;
    grid-template-columns: auto 1fr auto auto;
    grid-template-rows: 1fr 1fr;
    column-gap: ${theme.gutter};
    justify-content: start;
    text-align: left;
    align-items: center;

    @media screen and (max-width: 768px) {
      grid-template-columns: auto 1fr;
      grid-template-rows: auto 1fr auto auto;
      row-gap: ${theme.gutter};
    }
  `}
`

export const Avatar = styled.div`
  ${({ theme }) => css`
    height: 8rem;
    width: 8rem;
    border-radius: 50%;
    box-shadow: 0 0 0 0.1rem ${theme.colors.lightgray};
    overflow: hidden;
    grid-row: 1 / -1;
    position: relative;

    @media screen and (max-width: 768px) {
      grid-row: 1 / 2;
    }
  `}
`

export const Title = styled.p`
  ${({ theme }) => css`
    grid-column: 2 / 3;
    grid-row: 1 / 2;
    font-weight: ${theme.font.weight.bold};
    font-size: ${theme.font.size.xxlarge};
    color: ${theme.colors.darkPurple};

    @media screen and (max-width: 768px) {
      grid-column: 2 / 3;
      grid-row: 1 / 2;
    }
  `}
`

export const Description = styled.p`
  ${({ theme }) => css`
    grid-column: 2 / 4;
    grid-row: 2 / 3;
    font-size: ${theme.font.size.small};

    @media screen and (max-width: 768px) {
      grid-column: 1 / -1;
      grid-row: 2 / 3;
    }
  `}
`

const common = css`
  grid-row: 1 / 2;
  display: flex;
  align-items: center;
  font-size: 1.4rem;

  > svg {
    margin-right: 0.5rem;
  }

  @media screen and (max-width: 768px) {
    grid-column: 1 / -1;
    justify-self: end;
  }
`

export const Location = styled.p`
  grid-column: 3 / 4;
  ${common};

  @media screen and (max-width: 768px) {
    grid-row: 3 / 4;
  }
`

export const Website = styled.p`
  grid-column: 4 / 5;

  ${common};

  @media screen and (max-width: 768px) {
    grid-row: 4 / 5;
  }

  > a {
    text-decoration: none;
    color: inherit;
    display: inline-flex;
    align-items: center;
    column-gap: 0.2rem;

    > svg {
      font-size: 1rem;
    }
  }
`
