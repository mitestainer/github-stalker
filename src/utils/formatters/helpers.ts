export const formatDate = (dateString: string) => {
  return new Intl.DateTimeFormat('en-US', {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
  }).format(new Date(dateString))
}
