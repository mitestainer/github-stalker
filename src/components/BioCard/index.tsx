import Image from 'next/image'
import { FiGlobe, FiMapPin, FiExternalLink } from 'react-icons/fi'
import * as S from './styles'

type BioCardProps = {
  name: string
  avatar: string
  description?: string
  location?: string
  website?: string
}

const BioCard = ({
  name,
  avatar,
  description,
  location,
  website
}: BioCardProps) => {
  return (
    <S.Wrapper>
      <S.Avatar>
        <Image src={avatar} alt={name} layout="fill" objectFit="cover" />
      </S.Avatar>
      <S.Title>{name}</S.Title>
      {!!description && (
        <S.Description aria-label="description">{description}</S.Description>
      )}
      {!!location && (
        <S.Location>
          <FiMapPin aria-label="location" />
          {location}
        </S.Location>
      )}
      {!!website && (
        <S.Website>
          <FiGlobe aria-label={`${name}'s website`} />
          <a href={website} target="_blank" rel="noreferrer">
            <span>Website</span>
            <FiExternalLink />
          </a>
        </S.Website>
      )}
    </S.Wrapper>
  )
}

export default BioCard
