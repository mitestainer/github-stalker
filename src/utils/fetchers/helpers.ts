import { formatDate } from 'utils/formatters/helpers'

type RepoResponseProps = {
  id: number
  full_name: string
  description: string
  html_url: string
  watchers: number
  language: string
  updated_at: string
}

export const fetchRepos = async (handle: string, page: number) => {
  const response = await fetch(
    `https://api.github.com/orgs/${handle}/repos?sort=updated&direction=desc&page=${page}&per_page=12`
  )
  const data = await response.json()

  return data.map((repo: RepoResponseProps) => ({
    id: repo.id,
    name: repo.full_name,
    description: repo.description,
    url: repo.html_url,
    stars: repo.watchers,
    language: repo.language,
    updatedAt: formatDate(repo.updated_at)
  }))
}
