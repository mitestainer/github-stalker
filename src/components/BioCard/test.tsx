import { render, screen } from 'utils/tests/helpers'

import BioCard from '.'

const mockData = {
  name: 'Org Name',
  avatar: 'https://via.placeholder.com/150',
  description: 'The description text',
  location: 'World',
  website: 'www.google.com'
}

describe('<BioCard />', () => {
  it('should render the component with its data', () => {
    const { container } = render(<BioCard {...mockData} />)

    expect(screen.getByText(/org name/i)).toBeInTheDocument()
    expect(screen.getByRole('img', { name: mockData.name })).toHaveAttribute(
      'src',
      mockData.avatar
    )

    expect(screen.getByLabelText(/description/i)).toBeInTheDocument()
    expect(screen.getByText(/the description text/i)).toBeInTheDocument()

    expect(screen.getByLabelText(/location/i)).toBeInTheDocument()
    expect(screen.getByText(/world/i)).toBeInTheDocument()

    expect(screen.getByLabelText(/'s website$/i)).toBeInTheDocument()
    expect(screen.getByRole('link', { name: /website/i })).toHaveAttribute(
      'href',
      mockData.website
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should not render the description if not present', () => {
    render(<BioCard {...mockData} description="" />)

    expect(screen.queryByLabelText(/description/i)).not.toBeInTheDocument()
  })

  it('should not render the location if not present', () => {
    render(<BioCard {...mockData} location="" />)

    expect(screen.queryByLabelText(/location/i)).not.toBeInTheDocument()
  })

  it('should not render the website if not present', () => {
    render(<BioCard {...mockData} website="" />)

    expect(screen.queryByLabelText(/'s website$/i)).not.toBeInTheDocument()
  })
})
