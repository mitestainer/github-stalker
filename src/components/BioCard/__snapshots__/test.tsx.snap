// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BioCard /> should render the component with its data 1`] = `
.c0 {
  border-radius: 1rem;
  background-color: #fff;
  border: 1px solid #9387e1;
  padding: 2rem;
  display: grid;
  grid-template-columns: auto 1fr auto auto;
  grid-template-rows: 1fr 1fr;
  -webkit-column-gap: 2rem;
  column-gap: 2rem;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  text-align: left;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  height: 8rem;
  width: 8rem;
  border-radius: 50%;
  box-shadow: 0 0 0 0.1rem #dadae5;
  overflow: hidden;
  grid-row: 1 / -1;
  position: relative;
}

.c2 {
  grid-column: 2 / 3;
  grid-row: 1 / 2;
  font-weight: 600;
  font-size: 2.4rem;
  color: #5f49ca;
}

.c3 {
  grid-column: 2 / 4;
  grid-row: 2 / 3;
  font-size: 1.4rem;
}

.c4 {
  grid-column: 3 / 4;
  grid-row: 1 / 2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.4rem;
}

.c4 > svg {
  margin-right: 0.5rem;
}

.c5 {
  grid-column: 4 / 5;
  grid-row: 1 / 2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.4rem;
}

.c5 > svg {
  margin-right: 0.5rem;
}

.c5 > a {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-column-gap: 0.2rem;
  column-gap: 0.2rem;
}

.c5 > a > svg {
  font-size: 1rem;
}

@media screen and (max-width:768px) {
  .c0 {
    grid-template-columns: auto 1fr;
    grid-template-rows: auto 1fr auto auto;
    row-gap: 2rem;
  }
}

@media screen and (max-width:768px) {
  .c1 {
    grid-row: 1 / 2;
  }
}

@media screen and (max-width:768px) {
  .c2 {
    grid-column: 2 / 3;
    grid-row: 1 / 2;
  }
}

@media screen and (max-width:768px) {
  .c3 {
    grid-column: 1 / -1;
    grid-row: 2 / 3;
  }
}

@media screen and (max-width:768px) {
  .c4 {
    grid-column: 1 / -1;
    justify-self: end;
  }
}

@media screen and (max-width:768px) {
  .c4 {
    grid-row: 3 / 4;
  }
}

@media screen and (max-width:768px) {
  .c5 {
    grid-column: 1 / -1;
    justify-self: end;
  }
}

@media screen and (max-width:768px) {
  .c5 {
    grid-row: 4 / 5;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="Org Name"
      layout="fill"
      src="https://via.placeholder.com/150"
    />
  </div>
  <p
    class="c2"
  >
    Org Name
  </p>
  <p
    aria-label="description"
    class="c3"
  >
    The description text
  </p>
  <p
    class="c4"
  >
    <svg
      aria-label="location"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z"
      />
      <circle
        cx="12"
        cy="10"
        r="3"
      />
    </svg>
    World
  </p>
  <p
    class="c5"
  >
    <svg
      aria-label="Org Name's website"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="2"
        x2="22"
        y1="12"
        y2="12"
      />
      <path
        d="M12 2a15.3 15.3 0 0 1 4 10 15.3 15.3 0 0 1-4 10 15.3 15.3 0 0 1-4-10 15.3 15.3 0 0 1 4-10z"
      />
    </svg>
    <a
      href="www.google.com"
      rel="noreferrer"
      target="_blank"
    >
      <span>
        Website
      </span>
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
        />
        <polyline
          points="15 3 21 3 21 9"
        />
        <line
          x1="10"
          x2="21"
          y1="14"
          y2="3"
        />
      </svg>
    </a>
  </p>
</div>
`
